ocrd_kraken/recognize.py

```diff
@@ -0,0 +1,100 @@
+from ocrd import Processor
+from os.path import join
+from ocrd_utils import (
+    getLogger,
+    make_file_id,
+    assert_file_grp_cardinality,
+    bbox_from_polygon,
+    points_from_polygon,
+    points_from_bbox,
+    polygon_from_points,
+    bbox_from_points,
+    MIMETYPE_PAGE,
+)
+from ocrd_modelfactory import page_from_file
+from ocrd_models.ocrd_page import TextEquivType, WordType, GlyphType, CoordsType, to_xml
+
+from ocrd_kraken.config import OCRD_TOOL
+
+class KrakenRecognize(Processor):
+
+    def __init__(self, *args, **kwargs):
+        kwargs['ocrd_tool'] = OCRD_TOOL['tools']['ocrd-kraken-recognize']
+        kwargs['version'] = OCRD_TOOL['version']
+        super(KrakenRecognize, self).__init__(*args, **kwargs)
+
+    def process(self):
+        """
+        Recognize with kraken
+        """
+        from kraken.rpred import rpred
+        from kraken.lib.models import load_any
+        log = getLogger('processor.KrakenRecognize')
+        assert_file_grp_cardinality(self.input_file_grp, 1)
+        assert_file_grp_cardinality(self.output_file_grp, 1)
+        model_fname = self.resolve_resource(self.parameter['model'])
+        pad = self.parameter['pad']
+        bidi_reordering = self.parameter['bidi_reordering']
+        device = self.parameter['device']
+
+        log.info("loading model %s" % model_fname)
+        model = load_any(model_fname, device=device)
+
+        for n, input_file in enumerate(self.input_files):
+            page_id = input_file.pageId or input_file.ID
+            log.info("INPUT FILE %i / %s of %s", n, page_id, len(self.input_files))
+            pcgts = page_from_file(self.workspace.download_file(input_file))
+            self.add_metadata(pcgts)
+            page = pcgts.get_Page()
+            page_image, _, _ = self.workspace.image_from_page(page, page_id, feature_selector="binarized")
+
+            log.info("Converting PAGE to kraken 'bounds' format")
+            bounds = {'boxes': [], 'script_detection': True, 'text_direction': 'horizontal-lr'}
+            all_lines = page.get_AllTextLines()
+            for line in all_lines:
+                bounds['boxes'].append(bbox_from_points(line.get_Coords().points))
+
+            idx_line = 0
+            def _make_word(id_line, idx_word):
+                word = WordType(id='%s_word_%s' % (id_line, idx_word),
+                        Coords=CoordsType(points=''))
+                word.add_TextEquiv(TextEquivType(Unicode=''))
+                return word
+            for ocr_record in rpred(model, page_image, bounds, pad, bidi_reordering):
+                idx_word = 0
+                current_word = _make_word(all_lines[idx_line].id, idx_word)
+                idx_glyph = 0
+                for text, coords, conf in ocr_record:
+                    if text == ' ':
```

Comment:
This breaks if recognition starts with a whitespace character, because `current_word` is still the dummy from `make_word` without points. This should fix it:

```suggestion
                    if text == ' ' and idx_glyph > 0:
```

Reply:
No, wait. There could also be multiple whitespaces between words. So how about:
```suggestion
                    if text == ' ':
                        if idx_glyph == 0:
                            continue
```

Reply:
@bertsky That makes sense, I hadn't considered that edge case (and fear there might be more).

Just a reminder that you do have write permission to this PR, so feel free to directly commit if you so choose ;)

Reply:
> (and fear there might be more)

Yes, you should check for Unicode whitespace instead. The 'proper' way to do word segmentation would be to do splitting according to Unicode TR29 with something like the regex module, but words are a squishy concept in any case s somebody will always be unhappy.

Reply:
@mittagessen I don't think that applies here. Unicode TR29 uses a _semantic_ notion of word, so word boundaries tend to strip off punctuation. But PAGE's `Word` notion is purely _formal_, so words are simply tokens (only separated by whitespace). I doubt that OCR models will ever be trained with CR, LF, FF or NBS characters, so all we have to do here is collapse multiple SPACE.

Our concern here was merely with the correct "state machine" for the implementation (creating empty words and adding to them). There were [reports](https://gitter.im/OCR-D/Lobby?at=600a0073a2354e44ac9c703a) that this is still buggy yesterday – but I don't have the time right now to investigate, and I don't have the data to reproduce.

Reply:
> I doubt that OCR models will ever be trained with CR, LF, FF or NBS characters

But they are trained with the different width ones sometimes (mostly out of carelessness).

Reply:
> But they are trained with the different width ones sometimes (mostly out of carelessness).

Ah, good point. Chinese white space (and probably others) is an exception we should indeed support.